### ISSUE TITLE
error when using status.Conditions

### ISSUE THREAD

**Issue by raffaelespazzoli** [operator-framework/operator-sdk#2822]
## Bug Report
I'm getting this runtime error when updating a CR after migrating to using `status.Conditions` as per the docs:

```shell
"error":"EnforcingCRD.example.io \"example-enforcingcrd\" is invalid: status: Invalid value: \"array\": status in body must be of type object: \"array\""
```
the CR that I was trying to update is the following:

```json
{
   "kind":"EnforcingCRD",
   "apiVersion":"example.io/v1alpha1",
   "metadata":{
      "name":"example-enforcingcrd",
      "namespace":"test-enforcingcrd",
      "selfLink":"/apis/example.io/v1alpha1/namespaces/test-enforcingcrd/enforcingcrds/example-enforcingcrd",
      "uid":"4eaad05b-04b4-4fd8-9b5d-6f3cf162fc58",
      "resourceVersion":"8935334",
      "generation":1,
      "creationTimestamp":"2020-04-10T18:36:01Z",
      "annotations":{
         "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"example.io/v1alpha1\",\"kind\":\"EnforcingCRD\",\"metadata\":{\"annotations\":{},\"name\":\"example-enforcingcrd\",\"namespace\":\"test-enforcingcrd\"},\"spec\":{\"resources\":[{\"object\":{\"apiVersion\":\"v1\",\"data\":{\"ciao\":\"ciao\"},\"kind\":\"ConfigMap\",\"metadata\":{\"creationTimestamp\":\"2020-03-30T16:24:08Z\",\"name\":\"test-configmap\",\"namespace\":\"test-enforcingcrd\"}}},{\"object\":{\"apiVersion\":\"route.openshift.io/v1\",\"kind\":\"Route\",\"metadata\":{\"name\":\"test-route\",\"namespace\":\"test-enforcingcrd\"},\"spec\":{\"host\":\"grafana-istio-system.apps.cluster-4cac.sandbox456.opentlc.com\",\"tls\":{\"termination\":\"reencrypt\"},\"to\":{\"kind\":\"Service\",\"name\":\"grafana\",\"weight\":100},\"wildcardPolicy\":\"None\"}}}]}}\n"
      },
      "finalizers":[
         "enforcingcrd_controller"
      ]
   },
   "spec":{
      "resources":[
         {
            "object":{
               "apiVersion":"v1",
               "data":{
                  "ciao":"ciao"
               },
               "kind":"ConfigMap",
               "metadata":{
                  "creationTimestamp":"2020-03-30T16:24:08Z",
                  "name":"test-configmap",
                  "namespace":"test-enforcingcrd"
               }
            }
         },
         {
            "object":{
               "apiVersion":"route.openshift.io/v1",
               "kind":"Route",
               "metadata":{
                  "name":"test-route",
                  "namespace":"test-enforcingcrd"
               },
               "spec":{
                  "host":"grafana-istio-system.apps.cluster-4cac.sandbox456.opentlc.com",
                  "tls":{
                     "termination":"reencrypt"
                  },
                  "to":{
                     "kind":"Service",
                     "name":"grafana",
                     "weight":100
                  },
                  "wildcardPolicy":"None"
               }
            }
         }
      ]
   },
   "status":[
      {
         "type":"ReconcileSuccess",
         "status":"True",
         "reason":"Successful",
         "message":"Awaiting next reconciliation",
         "lastTransitionTime":"2020-04-10T18:36:27Z"
      }
   ]
}
```

Given the error message, and looking at the code, I sustepct the issue could be related to this:

```golang
// Conditions is a set of Condition instances.
//
// +kubebuilder:validation:Type=array
type Conditions map[ConditionType]Condition
```

The type is a map, but is mapped to an array, without specifying what the key would be.

**What did you do?**
see above

**What did you expect to see?**
I expected the update operation to be successful

**What did you see instead? Under which circumstances?**
the error aobve.

**Environment**
```
operator-sdk version
operator-sdk version: "v0.16.0", commit: "55f1446c5f472e7d8e308dcdf36d0d7fc44fc4fd", go version: "go1.13.8 linux/amd64"
```

* go version:
```
go version
go version go1.13.6 linux/amd64
```

* Kubernetes version information:

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.8-beta.0", GitCommit:"6c143d35bb11d74970e7bc0b6c45b6bfdffc0bd4", GitTreeState:"archive", BuildDate:"2020-01-02T00:00:00Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.2", GitCommit:"aa10b5b", GitTreeState:"clean", BuildDate:"2020-03-16T18:11:23Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

OpenShift

* Are you writing your operator in ansible, helm, or go?

go

**Possible Solution**
not sure, this seems just like a bug.

**Additional context**
none.

**Comment by camilamacedo86** [comment]
Could you please share the code snippet with the status and spec type definitions?

**Comment by raffaelespazzoli** [comment]
yes, here it is:
```golang
type MyCRDStatus struct {
	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html

	Conditions status.Conditions `json:"conditions"`
}
```
and here is how I set it:
```golang
	if reconcileStatusAware, updateStatus := (obj).(apis.ConditionsStatusAware); updateStatus {
		condition := status.Condition{
			Type:               "ReconcileError",
			LastTransitionTime: metav1.Now(),
			Message:            issue.Error(),
			Reason:             astatus.FailedReason,
			Status:             corev1.ConditionTrue,
		}
		reconcileStatusAware.SetReconcileStatus(status.NewConditions(condition))
```
here is the `SetReconcileStatus` method:
```golang
func (m *MyCRD) SetReconcileStatus(reconcileStatus status.Conditions) {
	m.Status.Conditions = reconcileStatus
}
```

**Comment by camilamacedo86** [comment]
Note that it provides a method for you to set the new conditional. See: https://godoc.org/github.com/operator-framework/operator-sdk/pkg/status#Conditions.SetCondition. 

Let us know if use the method provided solve your problem. However, anyway, I will create a POC as soon as possible and let you know.

**Comment by raffaelespazzoli** [comment]
@camilamacedo86 , thanks for looking at this also on the day of Easter.
with regards to your observation to add a new condition, I just always need to create a new one in this case. If you look at the json, I believe you'll find that it was created correctly. But OCP refuses to accept it likely for a marshaling issue on its side.
this [link](https://godoc.org/sigs.k8s.io/controller-tools/pkg/crd/markers#ListType) might be useful to solve the issue.

For example for pod's conditions:
```golang
...
	// +patchMergeKey=type
	// +patchStrategy=merge
	Conditions []PodCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
...
```

**Comment by joelanford** [comment]
@raffaelespazzoli This may be related to https://github.com/operator-framework/operator-sdk/pull/2739

Can you check to see if this issue is still occuring on `master`?

**Comment by joelanford** [comment]
If this is also solved by #2739, we may want to consider backporting to v0.16.1.

**Comment by raffaelespazzoli** [comment]
@joelanford What are the steps to check if it works with master? I am not very good at managing golang dependencies...

**Comment by joelanford** [comment]
in your project's go.mod, you should be able to do this:

```Go
require (
    github.com/operator-framework/operator-sdk master
)
```

Then run `go mod tidy` and `operator-sdk generate crds`

**Comment by joelanford** [comment]
Also, looking at the CR in your original message, you have:
```json
   "status":[
      {
         "type":"ReconcileSuccess",
         "status":"True",
         "reason":"Successful",
         "message":"Awaiting next reconciliation",
         "lastTransitionTime":"2020-04-10T18:36:27Z"
      }
   ]
```

Should that actually be the following?
```json
   "status":{
      "conditions":[
         {
            "type":"ReconcileSuccess",
            "status":"True",
            "reason":"Successful",
            "message":"Awaiting next reconciliation",
            "lastTransitionTime":"2020-04-10T18:36:27Z"
         }
      ]
   }
```

**Comment by joelanford** [comment]
I think there's a little bit of confusion that #2739 should clear up. The `Conditions` type is meant to be an abstraction, so you shouldn't have to worry about its underlying type when using it. 

In the first iteration, it was implemented as a map (for simplified set sematics) with custom marshal/unmarshal functions that would convert it to and from a JSON array. That actually all worked fairly well, except for when it came to using controller-tools to auto-generate the CRD validation. It turns out that using the `type=array` validation marker didn't actually do what we wanted it to do, so we ended up changing it to use a slice implementation instead.
